adoc/extensions/sycl_khr_group_interface.adoc

```diff
@@ -0,0 +1,479 @@
+
+_Returns_: An index space representing the collection of groups that includes
+this group, and which defines the range of valid [code]#id# values for this
+group.
+
+'''
+
+.[apidef]#GroupT::extents#
+[source,role=synopsis,id=api:khr-group-interface-common-group-extents]
+----
+constexpr extents_type extents() const noexcept;
+----
+
+_Returns_: The number of work-items in each dimension of the group.
+
+'''
+
+.[apidef]#GroupT::extent#
+[source,role=synopsis,id=api:khr-group-interface-common-group-extent]
+----
+constexpr index_type extent(rank_type r) const noexcept;
+----
+
+_Preconditions_: [code]#r < dimensions# is [code]#true#.
+
+_Returns_: The number of work-items in the specified dimension of the group.
+
+'''
+
+.[apidef]#GroupT::rank#
+[source,role=synopsis,id=api:khr-group-interface-common-group-rank]
+----
+static constexpr rank_type rank() noexcept;
+----
+
+_Effects_: Equivalent to [code]#return extents_type::rank();#.
+
+'''
+
+.[apidef]#GroupT::rank_dynamic#
+[source,role=synopsis,id=api:khr-group-interface-common-group-rank_dynamic]
+----
+static constexpr rank_type rank_dynamic() noexcept;
+----
+
+_Effects_: Equivalent to [code]#return extents_type::rank_dynamic();#.
+
+'''
+
+.[apidef]#GroupT::static_extent#
+[source,role=synopsis,id=api:khr-group-interface-common-group-static_extent]
+----
+static constexpr size_t static_extent(rank_type r) noexcept;
+----
+
+_Effects_: Equivalent to [code]#return extents_type::static_extent(r);#.
+
+'''
+
+.[apidef]#GroupT::size#
+[source,role=synopsis,id=api:common-group-size]
+----
+constexpr size_type size() const noexcept;
+----
+
+_Returns_: The total number of work-items in the group, equal to the product of
+the number of work-items in each dimension of the group.
+
+'''
+
+[[sec:khr-group-interface-common-non-member-funcs]]
+==== Non-member functions
+
+.[apidef]#get_item#
+[source,role=synopsis,id=api:common-group-get-item]
+----
+template <typename Group>
+work_item<Group> get_item(Group g) noexcept;
+----
+
+_Constraints_: Available only if `Group` is `work_group` or `sub_group`.
+
+_Returns_: A [code]#work_item# representing the calling work-item within group
+[code]#g#.
+
+'''
+
+.[apidef]#leader_of#
+[source,role=synopsis,id=api:common-group-leader_of]
+----
+template <typename Group>
+bool leader_of(Group g) noexcept;
+----
+
+_Returns_: [code]#true# if the calling work-item is the leader of group
+[code]#g#, and [code]#false# otherwise.
+
+_Remarks_: [code]#leader_of# returns [code]#true# for only one work-item in a
+group.
+The leader of the group is determined during construction of the group, and is
+invariant for the lifetime of the group.
+The leader of the group is guaranteed to be the work-item with index 0 within
+the group.
+
+[[sec:khr-group-interface-work_group]]
+== [code]#work_group# class
+
+The [code]#work_group# class template encapsulates all functionality required to
+represent a particular <<work-group>> within a kernel.
+It is not user-constructible.
```

Comment:
It is user constructible from a `group`.

Reply:
How about something like:

> The `work_group` class template encapsulates all functionality required to represent a specific work-group within a kernel.
> The set of work-items represented by an instance of the `work_group` class template is determined by the implementation, and there is subsequently no way for a user to construct arbitrary instances of the `work_group` class template.
> Instances of the `work_group` class template can only be acquired from a call to a standard SYCL function, or by converting an instance of the `group` class template.

I think this is the intent behind the "not user-constructible" wording in the SYCL 2020 specification, but I agree that intent is lost when we introduce something like a conversion constructor.

Reply:
Implemented in 3a6e6e2913b96b8b6674dbabb303e38230b27a69, with similar changes for `khr::work_group`, `khr::sub_group` and `khr::work_item`.